proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1378 @@
+There is not necessarily a problem with having some overlap between hooks and
+declarative features.
+
+An alternative approach would be to regard as illegitimate any use cases which
+treat `Cabal` as a packaging and distribution mechanism for executables, and on
+that basis, cease to provide copy hooks.  We do not follow this approach because
+it would significantly inconvenience maintainers of packages that rely on this
+behaviour (e.g. Agda and Darcs), for a relatively small reduction in complexity
+in `Cabal`.
+
+It is important that these copy hooks are also run when installing, as this
+fixes the inconsistency noted in [Cabal issue #709](https://github.com/haskell/cabal/issues/709).
+There is no separate notion of an "install hook", because "copy" and "install"
+are not distinct build phases.
+
+
+### Test and benchmark hooks
+
+Test and benchmark hooks are executed before or after the tests or benchmarks
+are compiled and run.  The main use cases for pre-test or pre-benchmark hooks
+are generating modules for use in tests or benchmarks; while technically this
+could be done in an earlier phase, if the generation step is expensive it makes
+sense to defer running it until the user actually requests tests or benchmarks.
+
+Another possible use cases is doctests, where:
+
+ * the package author wants `cabal test` to run doctests (so an external `cabal
+   doctest` command is not enough), e.g. because they use a different build tool
+   and need `./Setup test` to include doctests; and
+
+ * the `code-generators` feature is not suitable (e.g. because the package
+   includes multiple library components with incompatible GHC options).
+
+There are deliberately no package-level test or benchmark hooks, only
+component-level hooks. This is consistent with build hooks, and avoids
+introducing unnecessary synchronisation points if multiple packages are being
+tested in parallel.
+
+```haskell
+data TestComponentInputs
+  = TestComponentInputs
+  { additionalCommandLineArgs :: [String]
+  , localBuildInfo :: LocalBuildInfo
+  , testFlags :: TestFlags
+  , testSuite :: TestSuite
+  , componentLocalBuildInfo :: ComponentLocalBuildInfo
+  }
+
+data BenchmarkComponentInputs
+  = BenchmarkComponentInputs
+  { additionalCommandLineArgs :: [String]
+  , localBuildInfo :: LocalBuildInfo
+  , benchmarkFlags :: BenchmarkFlags
+  , benchmark :: Benchmark
+  , componentLocalBuildInfo :: ComponentLocalBuildInfo
+  }
+
+type TestComponentHook      = TestComponentInputs      -> IO ()
+type BenchmarkComponentHook = BenchmarkComponentInputs -> IO ()
+
+data TestHooks
+  = TestHooks
+  { preTestComponentHook  :: Maybe TestComponentHook
+  , postTestComponentHook :: Maybe TestComponentHook
+  }
+
+data BenchmarkHooks
+  = BenchmarkHooks
+  { preBenchComponentHook  :: Maybe BenchmarkComponentHook
+  , postBenchComponentHook :: Maybe BenchmarkComponentHook
+  }
+```
+
+
+### Clean hooks
+
+Clean hooks are useful if an earlier hook has generated some additional files
```

Comment:
We should follow the Nix practice where

1. The build system tells the arbitrary IO action where to put it's files.
2. Any IO action that does anything but put files in that spot is **wrong** and **unsupported**.

Clean hooks are bad, fixed filenames are also bad because they disallow fine-grained builds / not overwriting previous builds. This is what everyone expects coming from Bazel/Buck/Nix, and the legacy old Cabal/Hadrian/Make way of overwriting files because file names are normative is terrible and a huge drain on developer productivity.

Any architecture which doesn't support all intermediate files being named using content addresses is a bad architecture which can only be justified as an transitory step.